### DISCUSSION TITLE
Is it possible to optimize the ratio of multiple sums?

### DISCUSSION THREAD

**Discussion by adbok001** [google/or-tools#3236] [CP-SAT questions] [answered]
Hello,

I am trying to solve the following problem with the CP-SAT solver:

Imagine a hypothetical slot machine, that takes 10 items as input and randomly selects one of the possible outcomes as your prize.
The set of possible outcomes initially is empty and depends on the input items.
Every input item adds between 1 and 6 possible prizes with different values to the set of possible outcomes.
Every input item has four variables: weight, price, average value of the outcomes and number of possible outcomes.
My goal is to find the subset of 10 items with the highest profitabilit out of a huge set of items with different weights, prices and outcomes. 
Profitability is defined as the average value of the slot machine prizes divided by the total price for the 10 input items.
The average value of the slot machine prizes is the total value of all prizes divided by the number of possible outcomes.

In mathematical terms:

![math](https://user-images.githubusercontent.com/85845200/162652686-33b859ed-dc5d-4300-88db-3a1abb1a5258.PNG)

I know that this is probably NP-hard, so any approximate solution within reasonable time would be sufficient.

1. Is this possible with the CP-SAT solver? If so, what would be the most efficient way?

2. I am currently stuck at calculating the optimization function with `AddDivisionEquality` - as soon as I add it, the solver only instantly returns `INFEASIBLE`, even with a totally unrelated optimization function. Am I using `AddDivisionEquality` incorrectly?

My code:

```py
model = cp_model.CpModel()
x = {}
max_value = max(values)
max_outcomes = max(outcomes)
max_price = max(prices)

for i in data['all_items']:
    x[i] = model.NewBoolVar(f'x_{i}')

sum_outcomes = model.NewIntVar(0, max_outcomes * 10, 'sum_outcomes')
sum_prices = model.NewIntVar(0, max_price * 10, 'sum_prices')
prices_x_outcomes = model.NewIntVar(0, max_price * max_outcomes * 100, 'prices_x_outcomes')
sum_total = model.NewIntVar(0, max_outcomes * max_value * 10, 'sum_total')
profitability = model.NewIntVar(0, 100000, 'profitability')

model.Add(sum(x[i] * data['weights'][i] for i in data['all_items']) <= data['bin_capacity'])
model.Add(sum(x[i] for i in data['all_items']) == 10)
model.Add(sum_outcomes == sum(x[i] * data['outcomes'][i] for i in data['all_items']))
model.Add(sum_prices == sum(x[i] * data['prices'][i] for i in data['all_items']))
model.Add(sum_total == sum(x[i] * data['values'][i] * data['outcomes'][i] for i in data['all_items']))
model.AddMultiplicationEquality(prices_x_outcomes, [sum_outcomes, sum_prices])
### model.AddDivisionEquality(profitability, sum_total * 1000, prices_x_outcomes)

### model.Maximize(profitability)
```
Without the `AddDivisionEquality`, I can successfully optimize `prices_x_outcomes`, so the code until then seems to be fine - although it already seems to struggle a lot with 10k items. Sometimes the solution is found within a minute, sometimes it didn't find the solution after 10 minutes when I stopped it. Less items and it works reliably fast.

I'm happy to learn what I'm doing wrong here.

Thanks!

**Comment by lperron** [comment]
can you give a self contained code with a reasonable size (100 ?)

**Comment by adbok001** [comment]


**Reply by lperron** [reply]
you should replace the last print by     print(solver.ResponseStats())

**Comment by lperron** [comment, marked as answer]
The problem is that the divisor must be > 0, and the solver rejects the model as being invalid.

```
prices_x_outcomes = model.NewIntVar(1, max_price * max_outcomes * 100, 'prices_x_outcomes')
```

Now, you can set up logs, and enable parallelism

```
solver = cp_model.CpSolver()
solver.parameters.num_search_workers = 16
solver.parameters.log_search_progress = True
status = solver.Solve(model)
```

It finds a very good solution of 131 in 1.5s, and keeps working on the proof after that.

**Comment by adbok001** [comment]
@lperron Thanks, that makes sense. I thought about that first, but when I used an int (1) as denominator, the model still was seen as invalid.

I read about https://developers.google.com/optimization/cp/cp_tasks - is it also somehow possible to set a "dynamic" time limit, so that the solver terminates after X seconds without finding a better solution?

**Reply by vaidzelevich** [reply]
Hello. You can use a custom callback. You may see an example here https://github.com/stradivari96/ortools-utils/blob/master/ortools_utils/callbacks.py
